### ISSUE TITLE
Partial syncing of List trait

### ISSUE THREAD

**Issue by jvail** [jupyter-widgets/ipywidgets#2922]
Hi,

I have a List trait with items in my widget. The user can append items (dicts) to the list. I wonder if there is a way to sync only those items that have been added without sending the full list/state each time.
I thought about setting sync to False and send the new item with a custom message but then the model data in the browser is out of sync (I guess) and I can not save the complete widget state to the notebook file.
Any other ideas, suggestions?

Thank you
Jan

**Comment by cabreraalex** [comment]
I have a similar problem that would benefit from a solution to this:

I have a large list of image data that I can't afford to sync back and forth between the kernel and frontend. Is there some sort of Backbone collection for Widgets? I would love to query for a subset of a Traitlet list from the javascript frontend.

**Comment by ianhi** [comment]
I think the answer is yes but only with a decent effort. You would need to write your own serialization and deserialization methods that would only send the changes.

For image data compression can help a lot with constant syncing. In particular using zlib on the python side and [pako](https://www.npmjs.com/package/pako) on the typescript side. There is a library that implements this for numpy arrays already with the typescript serializers as a separate npm package so that you don't need to have users install multiple labextensions: https://github.com/vidartf/ipydatawidgets. I think that for syncing image data ipydatawidgets is currently the best approach.

There is some discussion on how to use it here: https://github.com/martinRenou/ipycanvas/issues/104 and and I have a single commit in which I started using it in my widget to sync image data that may be a helpful thing to look at: https://github.com/ianhi/ipysegment/commit/472353f76fb337484935260fbeecd9c4a9083ba8

**Comment by cabreraalex** [comment]
@ianhi you rock thanks so much for the help. I'll look into these and see if I can get something like that working!

**Comment by ianhi** [comment]
@cabreraalex no problem I think I was in your shoes just a few weeks ago. If what you're working on is open I'd love to see it. Also if you end up figuring out something slick for this please feel free to mention it on: https://github.com/ianhi/custom-ipywidget-howto/issues/4 and we can write it up into my unofficial documentation to help others in the future.

**Comment by martinRenou** [comment]
Leaving a link to spectate here: https://github.com/rmorshea/spectate.

Spectate is built on top of Traitlets and allows tracking changes on mutable data types. This is used in ipycytoscape for sending custom comm messages when lists content change.

**Comment by jvail** [comment]
Thank you for the hints. ipycytoscape & spectate look very interesting indeed!

**Comment by ianhi** [comment]
I definitely do not I fully understand how spectate works in ipycytoscape but I thought that it was just used to allow appending to lists or adding items to dictionaries cause a full traitlets sync. So the entire list still is synced via normal traitlets message it is just synced for `.append` events as well see https://github.com/QuantStack/ipycytoscape/blob/bb888cabba99bac4741f6783a0ba7db013e3f271/ipycytoscape/cytoscape.py#L113-L122

is that interpretation incorrect?

**Comment by martinRenou** [comment]
You are totally right indeed.

But we can imagine implementing a more refined solution where appending a value to a list would send a custom comm message to the front-end like `{ 'action': 'append', 'value': 3 }`. And updating a value in the list would send something like `{ 'action': 'update', 'index': 4, 'value': 3 }`.

That has been in my mind almost since I work with widgets. As updating mutable objects does not work, I find it very frustrating, and I am sure lots of users feel the same way.
I would love to see something like this included in core-ipywidgets.

**Comment by martinRenou** [comment]
> That would be implemented by creating custom serialization logic?

I actually don't think serialization would be involved, as serialization is for serializing an entire object into a JSON representation. I guess that would be some side-logic. Maybe the serialization would be involved at initialization and when a new value is assigned, and when methods like `append` are used we would use this side-logic that make use of custom comms.

> It seems like that could get complex rather quickly especially on the typescript side where different views could be making changes to a list and these changes might be conflicting.

Indeed. But we could maybe tackle the Python-side only first and consider this not supported JS-side for now.

> As an easier step maybe core-ipywidgets could incorporate spectate for lists, dicts and such?

I guess so

**Comment by cabreraalex** [comment]
Here was my hacky solution that works for my purpose, primarily I wanted to use the ipython kernel as a server and cache results in the frontend. It uses two traitlets, one that requests data and one that stores the response. When you want to get data you call the cache variable `get()` with a callback. It then either immediately calls the callback with the data, or sets the request traitlet and waits for data, then calls the callback.

```typescript
class ImageCache {
  model: WidgetModel;
  cache: Map<number, string> = new Map();

  callBack = (_: [number, string][]) => {};
  request: number[] = [];

  getImages(ids: number[], callBack: (_: [number, string][]) => void) {
    const requests = [];
    const imgs: string[] = [];
    ids.forEach(id => {
      const img = this.cache.get(id);
      if (!img) {
        requests.push(id);
      } else {
        imgs.push(img);
      }
    });
    if (requests.length !== 0) {
      this.callBack = callBack;
      this.request = ids;
      this.model.set('imageQueries', requests);
      this.model.save_changes();
    } else {
      callBack(zip(ids, imgs));
    }
  }


  imagesUpdated() {
    if (!this.model) return;
    const datas: DataView[] = this.model.get('imageDatas');
    const ids: number[] = this.model.get('imageQueries');
    datas.forEach((d, i) => {
      const blob: Blob = new Blob([d], {type: 'img/png'});
      this.cache.set(ids[i], URL.createObjectURL(blob));
    });
    this.callBack(zip(this.request, this.get(this.request)));
    this.callBack = () => {};
    this.request = [];
    this.model.set('imageQueries', []);
    this.model.set('imageDatas', []);
    this.model.save_changes();
  }

  get(ids: number[]): string[] {
    return ids.map(id => this.cache.get(id));
  }

  setModel(model: WidgetModel) {
    this.model = model;
    this.model.on('change:imageDatas', () =>this.imagesUpdated());
  }
}
```

**Comment by jvail** [comment]
> I would love to see something like this included in core-ipywidgets.
Yes, that would be great and I'd definitely prefer seeing such a feature implemented as part of ipywidgets rather than implementing my own.

But for now it seems then there is no way around working with custom messages (maybe on top of spectate). I have recently discovered that there is a "silent" mode for setting the model data in JS. This could be used for keeping the JS model in sync programmatically with custom messages without triggering any syncs. It seems a bit like a hack to me but re-sending the full state on each update isn't a solution either.

**Comment by ianhi** [comment]
I'm not sure that using `silent` is a good option. I think this will only have effects on JS side. In particular I think it will prevent other views from listening to the `change` event and so your views may get out of sync. Also from the backbonejs [docs](https://backbonejs.org/#Events-catalog):
>  if you'd like to prevent the event from being triggered, you may pass {silent: true} as an option. Note that this is rarely, perhaps even never, a good idea.

If the goal is to not send a large comm message back to python then I think the best approach is to be judicious with your usage of `this.model.save_changes()` or `this.model.touch()` as I believe that those are the only two methods that will trigger a sync back to python.

**Comment by jasongrout** [comment]
> `this.model.save_changes()` or `this.model.touch()` as I believe that those are the only two methods that will trigger a sync back to python.

IIRC, that is correct.

**Comment by jvail** [comment]
> I'm not sure that using `silent` is a good option.

Hm, yes. I was hoping that it has an effect on change detection as well but it doesn't. But not using model.touch() is not an option since there might be other properties that should sync. I noticed that backbone does not flag an array/obj as changed if the reference is unchanged. Not sure if I like the idea of manipulating the model state secretly but at least it seems this would not later not trigger any unwanted sync after calling touch().

**Comment by jvail** [comment]
What about backbone collections? I do not have great insight into the details of backbone or ipywidgets implementation but maybe it is feasible to implement a list of viewless models in python that would sync to a backbone collection in the front end..

**Comment by cabreraalex** [comment]
I think collections are the correct way to do this from the Backbone perspective, but I'm not familiar enough with the IPywidget implementation to know if that would be an easy change/update. Keeping track of hundreds or thousands of models might be really slow? Maybe extending a traitlet to represent a backbone collection?

**Comment by ianhi** [comment]
Something I found to be fairly illuminating was reading the source of the widget object you inherit from:
https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/base/src/widget.ts

and the source of the set method: https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/base/src/backbone-patch.ts

Since the backbone set method is already being patched maybe it could be patched further to allow an optional argument that would prevent those changes being synced back to python? Or `save_changes` could accept a list of variables to ignore for that save. This would at least allow changing the model as desired on the typescript side and only syncing when desired

**Comment by ianhi** [comment]
Another place where the idea of sending a diff of the change was brought up: https://github.com/jupyter-widgets/traittypes/pull/36#issuecomment-502217109

**Comment by jvail** [comment]
@ianhi Thank you for posting those links. Have not found any time to dig through the sources. Good to read that there is some work in progress already.
